sextant/src/npk.rs

```diff
@@ -286,72 +286,86 @@ fn gen_pseudo_files(manifest: &Manifest) -> Result<Vec<(&str, u32)>> {
+
     let mut hash_tree = vec![0_u8; tree_size];
-    let hash_src_offset = 0;
-    let mut hash_src_size = image_size;
     let mut level_num = 0;
-    let mut reader = BufReader::new(image);
-    let mut level_output: Vec<u8> = vec![];
+    let mut level_size = image_size;
+    let mut level_hashes: Vec<[u8; SHA256_SIZE]> = vec![];
+
+    if image_size % BLOCK_SIZE as u64 != 0 {
+        return Err(anyhow!(
+            "Failed to generate verity has tree. The image size {} is not a multiple of the block size {}",
+            image_size,
+            BLOCK_SIZE
+        ));
+    }
+
+    loop {
+        level_hashes.clear();
+        let mut rem_size = level_size;
 
-    while hash_src_size > block_size {
-        let mut level_output_list: Vec<[u8; SHA256_SIZE]> = vec![];
-        let mut remaining = hash_src_size;
-        while remaining > 0 {
+        while rem_size > 0 {
+            // 1. Choose a random salt (hexadecimal encoding).
             let mut sha256 = Sha256::new();
             sha256.update(salt);
 
-            let data_len;
+            // "2. Unsparse your system image into 4k blocks."
+            // "3. For each block, get its (salted) SHA256 hash."
             if level_num == 0 {
-                let offset = hash_src_offset + hash_src_size - remaining;
-                data_len = std::cmp::min(remaining, block_size);
-                let mut data = vec![0_u8; data_len as usize];
-                reader.seek(Start(offset))?;
-                reader.read_exact(&mut data)?;
+                // hash block of original file
+                let offset = level_size - rem_size;
+                let mut data = vec![0_u8; BLOCK_SIZE];
+                let mut image_reader = BufReader::new(image);
+                image_reader.seek(Start(offset))?;
+                image_reader.read_exact(&mut data)?;
                 sha256.update(&data);
             } else {
-                let offset =
-                    hash_level_offsets[level_num - 1] + hash_src_size as usize - remaining as usize;
-                data_len = block_size;
-                sha256.update(&hash_tree[offset..offset + data_len as usize]);
+                // hash block of previous level
+                let offset = level_offsets[level_num - 1] + level_size as usize - rem_size as usize;
+                sha256.update(&hash_tree[offset..offset + BLOCK_SIZE]);
             }
 
-            remaining -= data_len;
-            if data_len < block_size {
-                let zeros = vec![0_u8; (block_size - data_len) as usize];
-                sha256.update(zeros);
-            }
-            level_output_list.push(sha256.finalize().into());
+            rem_size -= BLOCK_SIZE as u64;
+            level_hashes.push(sha256.finalize().into());
+        }
+
+        // "7. Repeat steps 2-6 using the previous level as the source for the next until you have only a single hash."
+        if level_hashes.len() == 1 {
+            break;
         }
```

Comment:
Why is step 7 before step 4?

Reply:
This comes straigt from the algorithm description [1] in which steps 2-6 describe a loop and step 7 describes the condition to break out of it. I'll remove the leading '7.' to avoid the confusion.

[1] https://source.android.com/security/verifiedboot/dm-verity#hash-tree